lib/datadog/core/telemetry/metrics_collection.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+require_relative 'event'
+require_relative 'metric'
+
+module Datadog
+  module Core
+    module Telemetry
+      # MetricsCollection is a thread-safe collection of metrics per namespace
+      class MetricsCollection
+        attr_reader :namespace, :interval
+
+        def initialize(namespace, aggregation_interval:)
+          @namespace = namespace
+          @interval = aggregation_interval
+
+          @mutex = Mutex.new
+
+          @metrics = {}
+          @distributions = {}
+        end
+
+        def inc(metric_name, value, tags: {}, common: true)
+          metric = Metric::Count.new(metric_name, tags: tags, common: common)
+
+          @mutex.synchronize do
+            metric = fetch_or_add_metric(metric, @metrics)
```

Comment:
Nitpick: It looks weird to me passing instance variable to its private method, which already has access to the state of the object. My guess is that you want to make `fetch_or_add_metric` generic for both `@metrics` and `distributions` 

However, I would actually prefer separating these two.

```ruby
 @mutex.synchronize do
  (@metrics[metric.id] ||= metric).tap { |m| m.inc(value) }
end
```
It also put them closer to `@mutex` preventing usage without it.

I would also be fine if you pass instance variable to the private method and refactor `fetch_or_add_metric` to include the `@mutex.sychronization`

Reply:
I don't have strong opinion on this, your approach works for me too

Reply:
this one looks very clean, I will try it

Reply:
Ah wait, there's a bug in my code! It should be

```ruby
def fetch_or_add_metric(metric)
  @mutex.synchronize do
    yield (collection[metric.id] ||= metric)
  end
  nil
end

Reply:
>...maybe this was what Tony had in mind?

You read my mind.